# Cell 1: đọc file relations và xem cột: mỗi nghề đi kèm những skill gì

Mục đích:

- mở file occupationSkillRelations_en.csv

- kiểm tra file này có những cột gì

- xác định cột nào dùng để nối:

    - occupationLabel

    - skillLabel


* Đọc file quan hệ occupation-skill của ESCO để kiểm tra cấu trúc dữ liệu và xác định các cột cần dùng khi lọc.

In [2]:
import pandas as pd

df_rel = pd.read_csv(
    "/Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/dataset/ESCO_taxonomy/esco_keep/occupationSkillRelations_en.csv"
)

print(df_rel.columns.tolist())
print(df_rel.shape)
df_rel.head()

['occupationUri', 'occupationLabel', 'relationType', 'skillType', 'skillUri', 'skillLabel', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9']
(126051, 10)


,occupationUri,occupationLabel,relationType,skillType,skillUri,skillLabel,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9
0,http://data.europa.eu/esco/occupation/00030d09...,technical director,essential,knowledge,http://data.europa.eu/esco/skill/fed5b267-73fa...,theatre techniques,NaN,NaN,NaN,NaN
1,http://data.europa.eu/esco/occupation/00030d09...,technical director,essential,skill/competence,http://data.europa.eu/esco/skill/05bc7677-5a64...,organise rehearsals,NaN,NaN,NaN,NaN
2,http://data.europa.eu/esco/occupation/00030d09...,technical director,essential,skill/competence,http://data.europa.eu/esco/skill/271a36a0-bc7a...,write risk assessment on performing arts produ...,NaN,NaN,NaN,NaN
3,http://data.europa.eu/esco/occupation/00030d09...,technical director,essential,skill/competence,http://data.europa.eu/esco/skill/47ed1d37-971b...,coordinate with creative departments,NaN,NaN,NaN,NaN
4,http://data.europa.eu/esco/occupation/00030d09...,technical director,essential,skill/competence,http://data.europa.eu/esco/skill/591dd514-735b...,adapt to artists' creative demands,NaN,NaN,NaN,NaN


Mục đích:

- nạp 2 nguồn dữ liệu đầu vào:

    - file nghề IT đã gộp

    - file quan hệ occupation-skill của ESCO

* Nạp tập nghề IT đã lọc và bảng quan hệ occupation-skill để chuẩn bị cho bước lọc quan hệ theo nghề IT.

In [3]:
import pandas as pd
import os

# ===== 1. Load merged IT occupations =====
df_occ_all = pd.read_csv(
    "/Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/notebook/ESCO_taxonomy/notebook_clean/06_esco_it_all_occupations.csv"
)

# ===== 2. Load ESCO occupation-skill relations =====
df_rel = pd.read_csv(
    "/Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/dataset/ESCO_taxonomy/esco_keep/occupationSkillRelations_en.csv"
)

print("Relations columns:", df_rel.columns.tolist())
print("Relations shape:", df_rel.shape)

Relations columns: ['occupationUri', 'occupationLabel', 'relationType', 'skillType', 'skillUri', 'skillLabel', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9']
Relations shape: (126051, 10)


# Cell 2 chuẩn hơn: lọc relations theo tập nghề IT

In [ ]:
import pandas as pd
import os

# ===== 1. Load merged IT occupations =====
df_occ_all = pd.read_csv(
    "/Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/notebook/ESCO_taxonomy/notebook_clean/06_esco_it_all_occupations.csv"
)

# ===== 2. Load ESCO occupation-skill relations =====
df_rel = pd.read_csv(
    "/Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/dataset/ESCO_taxonomy/esco_keep/occupationSkillRelations_en.csv"
)

# ===== 3. Drop unnamed columns =====
df_rel = df_rel.loc[:, ~df_rel.columns.str.contains("^Unnamed")].copy()

# ===== 4. Clean labels =====
df_occ_all["preferredLabel_clean"] = (
    df_occ_all["preferredLabel"]
    .fillna("")
    .astype(str)
    .str.strip()
    .str.lower()
)

df_rel["occupationLabel_clean"] = (
    df_rel["occupationLabel"]
    .fillna("")
    .astype(str)
    .str.strip()
    .str.lower()
)

df_rel["skillLabel_clean"] = (
    df_rel["skillLabel"]
    .fillna("")
    .astype(str)
    .str.strip()
    .str.lower()
)

# ===== 5. Create IT occupation set =====
it_occ_set = set(df_occ_all["preferredLabel_clean"])

# ===== 6. Filter only relations whose occupation is in IT set =====
df_rel_it = df_rel[df_rel["occupationLabel_clean"].isin(it_occ_set)].copy()

# Nếu occupationLabel_clean trong relation
# có nằm trong danh sách occupation IT đã lọc không?

# Có → giữ relation đó
# Không → bỏ

print("Total IT occupation-skill relations:", len(df_rel_it))
print("Unique IT occupations in relations:", df_rel_it["occupationLabel_clean"].nunique())
print("Unique IT skills in relations:", df_rel_it["skillLabel_clean"].nunique())

df_rel_it.head()

# Kết quả sau khi lọc relations theo nghề IT

Total IT occupation-skill relations: 3275
Unique IT occupations in relations: 54
Unique IT skills in relations: 1172


,occupationUri,occupationLabel,relationType,skillType,skillUri,skillLabel,occupationLabel_clean,skillLabel_clean
8948,http://data.europa.eu/esco/occupation/10469d70...,embedded system designer,essential,knowledge,http://data.europa.eu/esco/skill/09f2f811-a3fb...,systems development life-cycle,embedded system designer,systems development life-cycle
8949,http://data.europa.eu/esco/occupation/10469d70...,embedded system designer,essential,knowledge,http://data.europa.eu/esco/skill/2180bd8c-86de...,embedded systems,embedded system designer,embedded systems
8950,http://data.europa.eu/esco/occupation/10469d70...,embedded system designer,essential,knowledge,http://data.europa.eu/esco/skill/397da142-ab35...,digital systems,embedded system designer,digital systems
8951,http://data.europa.eu/esco/occupation/10469d70...,embedded system designer,essential,knowledge,http://data.europa.eu/esco/skill/3c42d13b-d6e4...,engineering control theory,embedded system designer,engineering control theory
8952,http://data.europa.eu/esco/occupation/10469d70...,embedded system designer,essential,knowledge,http://data.europa.eu/esco/skill/6e4f75b4-c60f...,ICT communications protocols,embedded system designer,ict communications protocols


# Kết quả sau khi lọc relations theo nghề IT

Ở đây mày thấy:

- Total IT occupation-skill relations: 3275 => Đây là tổng số dòng quan hệ nghề - skill sau khi mày lọc chỉ còn nghề IT.

- Unique IT occupations in relations: 54 => Đây là số nghề IT khác nhau xuất hiện trong bảng relation sau lọc.

- Unique IT skills in relations: 1172 => Đây là số skill khác nhau đi kèm với các nghề IT sau lọc.

Mục đích:

- chỉ giữ các dòng mà occupation thuộc tập nghề IT

- thống kê nhanh số lượng quan hệ, số nghề, số skill sau lọc

* Lọc bảng quan hệ để chỉ giữ các dòng có occupation thuộc tập nghề IT, sau đó thống kê số lượng quan hệ, số nghề và số kỹ năng thu được.

# Xem sample dữ liệu sau lọc

Ví dụ như:

- embedded system designer

- skill đi kèm của nó

Mục đích:

- kiểm tra nhanh bằng mắt xem dữ liệu sau lọc có hợp lý không

- xác nhận occupation IT đang đi với các skill phù hợp

* Kiểm tra mẫu dữ liệu sau lọc để xác nhận các occupation IT được ghép với các skill phù hợp.

# Gắn group core / extended vào relations

Kết quả:

- core: 2847

- extended: 428

Mục đích:

- nối ngược từ file nghề IT tổng để biết mỗi quan hệ thuộc nhóm nào

- giúp tách riêng relations của core và extended nếu cần

Gắn nhãn core/extended cho từng dòng quan hệ occupation-skill dựa trên nhóm nghề đã phân loại trước đó vì: File occupations đã có nhãn group, còn file occupation-skill relations 

thì chưa có, nên cần map lại nhãn core/extended từ occupation sang relation.

In [ ]:
# ===== 7. Map group from occupation file =====
occ_group_map = dict(
    zip(df_occ_all["preferredLabel_clean"], df_occ_all["group"])
)

df_rel_it["group"] = df_rel_it["occupationLabel_clean"].map(occ_group_map)

print(df_rel_it["group"].value_counts(dropna=False))
df_rel_it[["occupationLabel", "skillLabel", "group"]].head(20)

# Đây là số dòng relation thuộc từng nhóm.

# Tức là trong 3275 dòng tổng:

# 2847 dòng thuộc nhóm core

# 428 dòng thuộc nhóm extended

# Hiểu đơn giản:

# các nghề core kéo theo tổng cộng 2847 quan hệ nghề-skill

# các nghề extended kéo theo tổng cộng 428 quan hệ nghề-skill


group
core        2847
extended     428
Name: count, dtype: int64


,occupationLabel,skillLabel,group
8948,embedded system designer,systems development life-cycle,core
8949,embedded system designer,embedded systems,core
8950,embedded system designer,digital systems,core
8951,embedded system designer,engineering control theory,core
8952,embedded system designer,ICT communications protocols,core
8953,embedded system designer,task algorithmisation,core
8954,embedded system designer,tools for software configuration management,core
8955,embedded system designer,real-time computing,core
8956,embedded system designer,signal processing,core
8957,embedded system designer,system design,core


# Export file sạch

In [6]:
output_dir = "/Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/notebook/ESCO_taxonomy/notebook_clean"
os.makedirs(output_dir, exist_ok=True)

output_path = os.path.join(output_dir, "07_esco_it_occupation_skill_relations.csv")
df_rel_it.to_csv(output_path, index=False)

print(f"Saved IT occupation-skill relations to: {output_path}")

Saved IT occupation-skill relations to: /Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/notebook/ESCO_taxonomy/notebook_clean/07_esco_it_occupation_skill_relations.csv
